### ISSUE TITLE
Support hot-reloading of certificates

### ISSUE THREAD

**Issue by jceb** [kanidm/kanidm#1271]
### I did this

I'd like to operate kanidm in a kubernetes deployment with cert-manager for generating the certificates. I tried to trigger a hot-reload of kanidm's LDAPS and HTTPS services by generating a new certificate on top of the existing files, I also tried to `touch chain.pem key.pem`  but neither of the two attempts yielded a reload.

### I expected the following

Since cert-manager and other tools automatically generate and update certificates, it would be great if kanidm would watch certificate and key files (attention, files might be symlinks to other symlinks that eventually point to a regular file) and trigger an automatic reload when they change.

### Kanidm version (and git commit)

kanidm 1.1.0-alpha.10
2022-12-16T10:00:54.028061Z DEBUG kanidm: Using 16 worker threads

kanidmd 1.1.0-alpha.11-dev (in docker)

### Operating System / Version

Linux torch 6.0.12-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 08 Dec 2022 11:03:38 +0000 x86_64 GNU/Linux

**Comment by Firstyear** [comment]
@jceb I actually was thinking about this the other day. I think I will not add support for this, because there are many reasons to need to restart the server anyway. Consider security patches, general updates, cluster node failures. 

The other main reason is this feature is a non trivial amount of work, and I'd rather spend that time implementing replication (once I finish the current set of IDM sync features that have been requested). 

I think we should aim to make kanidm reliable as a topology with replication, rather than trying to keep a single node running "for as long as possible".

**Comment by yaleman** [comment]
In the k8s case it should be relatively easy to solve this - put a long-lived cert using the selfsigned CA on the pod, and your short-lived cert on the ingress controller. The ingress should trust the CA on the internal comms, thus maintaining security.

**Comment by jceb** [comment]
@yaleman I'm not super happy with such workarounds. My preference is to have kanidm play nicely with Kubernetes and the patterns that already exist in this environment. The long-lived certificate generates the same issue again a few years down the road. Since there are good tools for managing certificates, I'd try to integrate with them.

@Firstyear I agree that a reliable topology should be the desired state. The right point in time for restarting or reloading a service still need to be identified. The manual intervention by an administrator is always an option, which options do you see for automating restarts/reloads?

In Kubernetes a change of the deployment will trigger a hard restart of the pod/service. However, a change in referenced configurations/ConfigMaps and secrets doesn't lead to a restart of the service. An administrator would have to manually trigger a restart if the service doesn't notice the change. In my experience there are good mechanisms in place for triggering restarts whenever a configuration changes. Especially for certificate secrets the handling is different since certificate issuance and key rotation are usually not done manually by an administrator. Instead tools like cert-manager or Hashicorp Vault are used. These tools will update the the related secret but they won't alter the deployment and therefore not trigger a forced restart.

**Comment by yaleman** [comment]
> The long-lived certificate generates the same issue again a few years down the road.

It only has to last as long as the pod, which if you have them a long time, is a process smell in itself. It's not perfect, but certificates can be issued by vault/cert-manager on pod creation and only need to secure traffic between pod and ingress.

> However, a change in referenced configurations/ConfigMaps and secrets doesn't lead to a restart of the service.

Not by default, but controllers like [Reloader](https://github.com/stakater/Reloader) allow for this to be managed trivially. All of this can be managed through existing automations, lots faster than the rather messy rewrite of the Kanidm server can likely be done. Anyone's welcome to implement it, though 😃 

Note that `tide-rs` is not really being maintained and there has been small discussions around ways of replacing its use in Kanidm, due to other issues that it has, so further investment in it aren't high priority when compared to other feature dev.

**Comment by Firstyear** [comment]
@jceb At the end of the day, the issue comes down to time. Do I spend a month or two working on a feature that keeps a single node up for a "little bit longer"? Or do I work on replication?

No matter how you look at it, you'll need to restart your kanidm nodes at some point, we can't keep them "alive for ever". This feature only delays the inevitable restart by a few days if not weeks. 

It's better to build your infra to handle the restart when needed. And as mentioned, by NOT doing this feature, that means we can implement replication sooner which also resolves the problem more broadly. 

So I think that in my view, while I like the *idea* of this feature, I think reality is that it won't be useful for very long, and it will detract from our time to implement more important things.

**Comment by yaleman** [comment]
Reopening this as there's ways/reasons to do it

**Comment by yaleman** [comment]
axum-server supports reloading the config as part of a niceness too as of 0.6.0 - https://github.com/programatik29/axum-server/releases/tag/v0.6.0

**Comment by jinnatar** [comment]
As someone now attempting to run the certs from HashiCorp Vault I'll add my 2c. My goal is to have short lived (say 24h) certs from kanidmd to ingress rendered from Vault via the consul-template `pkiCert` function. [docs](https://github.com/hashicorp/consul-template/blob/main/docs/templating-language.md#pkicert). This will ensure the cert files on disk remain valid, so all that's missing is to communicate to kanidmd to reload the certs. The easy options available to me right now in Nomad is to send a signal to the container pid 1, or to restart the task. (There's the option also of running an arbitrary script, but that's not useful in this scenario. [Full change_mode docs](https://developer.hashicorp.com/nomad/docs/job-specification/template#change_mode))

So right now I'll just instruct it to restart the task, and allowing that every 24h or so is not too bad. There's very little smarts in that restart that I'm aware of though, so assuming I launch say 5 instances of replicated kanidmd they'll come up in series, render new certs & get healthy etc less than a minute apart and end up with certs with expirations close to eachother. I'm assuming this will mean the restarts in turn will happen relatively close to eachother. All of this can be worked around and improved with smarter orchestration, BUT from an operators perspective it would still be safer if instead of a restart I could trigger a reload of the certs.

In summary: restarting the task to reload certs is "ok", but reloading them on say SIGUSR1 would take away complexity from making sure the restarts don't all align on the same damn moment. :-)
